### PR TITLE
XRC: add GetBitmapBundle function

### PR DESCRIPTION
**Pull request by kosh543**
Ability to create wxBitmapBundle in XRC:
`<bitmap>new.png;new_2x.png</bitmap>`
`<bitmap default_size="32,32">new.svg</bitmap>`

### CODE REVIEW

File: docs/doxygen/overviews/xrc_format.h [hunk 1]
```diff
@@ -363,6 +364,19 @@ percent-encoded, e.g. here is the correct way to specify a bitmap with the path
 Bitmap file paths can include environment variables that are expanded if
 wxXRC_USE_ENVVARS was passed to the wxXmlResource constructor.
 
+It is possible to specify the multi-resolution bitmap by a set of bitmaps or SVG file,
+which are mutually exclusive. The set of bitmaps should contain few relative URLs of
```
Comment (vadz) [inline docs/doxygen/overviews/xrc_format.h:368]:
```suggestion
It is possible to specify the multi-resolution bitmap by a set of bitmaps or an SVG file,
which are mutually exclusive. The set of bitmaps should contain one or more relative URLs of
```

File: docs/doxygen/overviews/xrc_format.h [hunk 2]
```diff
@@ -363,6 +364,19 @@ percent-encoded, e.g. here is the correct way to specify a bitmap with the path
...
+a bitmap, separated by @c ';'.
+Example with a set of bitmaps:
+@code
+<bitmap>new.png;new_2x.png</bitmap>
+@endcode
+While using SVG file you also should specify @c default_size attribute,
+because usually SVG file doesn't have it:
```
Comment (vadz) [inline docs/doxygen/overviews/xrc_format.h:375]: I'd like to rephrase this as
```suggestion
When using SVG file you must also specify @c default_size attribute
(even if the size is specified in SVG file, it may be different from the size needed here):
```

File: docs/doxygen/overviews/xrc_format.h [hunk 3]
```diff
@@ -363,6 +364,19 @@ percent-encoded, e.g. here is the correct way to specify a bitmap with the path
 Bitmap file paths can include environment variables that are expanded if
 wxXRC_USE_ENVVARS was passed to the wxXmlResource constructor.
 
+It is possible to specify the multi-resolution bitmap by a set of bitmaps or SVG file,
+which are mutually exclusive. The set of bitmaps should contain few relative URLs of
+a bitmap, separated by @c ';'.
+Example with a set of bitmaps:
+@code
+<bitmap>new.png;new_2x.png</bitmap>
+@endcode
```
Comment (vadz) [inline docs/doxygen/overviews/xrc_format.h:373]: I think this is not detailed enough, let's explain this in more details:
```suggestion
For example, to specify two bitmaps, to be used in standard and 200% DPI scaling respectively, you could write:
@code
<bitmap>new.png;new_2x.png</bitmap>
@endcode

Here the first bitmap is special, as its size determines the logical size of the bitmap. In other words, this bitmap is the one used when DPI scaling is not in effect. Any subsequent bitmaps can come in any order and will be used when the DPI scaling factor is equal, or at least close, to the ratio of their size to the size of the first bitmap. Using @c _2x naming convention here is common, but @e not required, the names of the bitmaps can be arbitrary, e.g.
@code
<bitmap>new_32x32.png;new_64x64.png</bitmap>
@endcode
would work just as well.
```

File: src/xrc/xmlres.cpp [hunk 4]
```diff
@@ -2182,29 +2302,26 @@ void XRCConvertFromDLU(wxWindow* w, T& value)
...
 
-// Helper for parsing values (of type T, for which XRCConvertFromAbsValue() and
-// XRCConvertFromDLU() functions must be defined) which can be expressed either
-// in pixels or dialog units.
 template <typename T>
 T
-ParseValueInPixels(wxXmlResourceHandlerImpl* impl,
+ParseStringInPixels(wxXmlResourceHandlerImpl* impl,
                    const wxString& param,
+                   const wxString& str,
```
Comment (vadz) [inline src/xrc/xmlres.cpp:2309]: Pretty minor, but why rename this to `str` from `s`? The name is not that much better, but results in several diff hunks below, that would be avoided if we kept the old name.

File: src/xrc/xmlres.cpp [hunk 5]
```diff
@@ -1852,6 +1852,52 @@ bool GetStockArtAttrs(const wxXmlNode *paramNode,
...
+            nodeName,
+            wxString::Format("cannot create bitmap from \"%s\"", path)
+        );
+        return wxNullBitmap;
+    }
+    if (!(size == wxDefaultSize)) img.Rescale(size.x, size.y);
+    return wxBitmap(img);
+}
+
+// forward declaration
```
Comment (vadz) [inline src/xrc/xmlres.cpp:1892]: I think the comment in front of `ParseValueInPixels()` should be moved either here or before this function definition below, and the other function comment updated to just refer to this one.

The point is that it would be more helpful to have a comment for this function, which is where the real work is being done and which occurs first in the sources, rather than on the other one.

File: src/xrc/xmlres.cpp [hunk 6]
```diff
@@ -1852,6 +1852,52 @@ bool GetStockArtAttrs(const wxXmlNode *paramNode,
     return false;
 }
 
+// Load a bitmap from a file.
+wxBitmap LoadBitmapFromFile(wxXmlResourceHandlerImpl* impl,
```
Comment (vadz) [inline src/xrc/xmlres.cpp:1856]: This is not necessarily a file, is it? It's `wxFileSystem` element, so maybe rename it like this
```suggestion
wxBitmap LoadBitmapFromFS(wxXmlResourceHandlerImpl* impl,
```
?

File: src/xrc/xmlres.cpp [hunk 7]
```diff
@@ -1892,36 +1938,110 @@ wxBitmap wxXmlResourceHandlerImpl::GetBitmap(const wxXmlNode* node,
...
+                          art_id, art_client) )
+    {
+        wxBitmapBundle stockArt(wxArtProvider::GetBitmapBundle(art_id, art_client, size));
+        if ( stockArt.IsOk() )
+            return stockArt;
+    }
+
+    wxBitmapBundle bitmapBundle;
+    wxString paramValue = GetParamValue(node);
+    bool isBitmaps = paramValue.Contains(";");
```
Comment (vadz) [inline src/xrc/xmlres.cpp:1971]: I'd like to rename the variable to make it more clear what it is about and make it const to show that it never changes below:
```suggestion
    const bool hasSeveralBitmaps = paramValue.Contains(";");
```

File: src/xrc/xmlres.cpp [hunk 8]
```diff
@@ -1892,36 +1938,110 @@ wxBitmap wxXmlResourceHandlerImpl::GetBitmap(const wxXmlNode* node,
...
+    bool isSVG = paramValue.EndsWith(".svg");
```
Comment (vadz) [inline src/xrc/xmlres.cpp:1972]: Also
```suggestion
    const bool isSVG = paramValue.EndsWith(".svg");
```

File: src/xrc/xmlres.cpp [hunk 9]
```diff
@@ -1892,36 +1938,110 @@ wxBitmap wxXmlResourceHandlerImpl::GetBitmap(const wxXmlNode* node,
...
+    if ( isBitmaps && isSVG )
     {
         ReportParamError
         (
-            node->GetName(),
-            wxString::Format("cannot create bitmap from \"%s\"", name)
+            param,
+            "may contain either one svg file or a list of files separated by ';'"
```
Comment (vadz) [inline src/xrc/xmlres.cpp:1978]:
```suggestion
            "may contain either one SVG file or a list of files separated by ';'"
```

File: src/xrc/xmlres.cpp [hunk 10]
```diff
@@ -1892,36 +1938,110 @@ wxBitmap wxXmlResourceHandlerImpl::GetBitmap(const wxXmlNode* node,
...
+                "SVG bitmaps are not supported in this build of the library"
+            );
+#endif // wxHAS_SVG/!wxHAS_SVG
+        }
+    }
+    else
+    {
+        // it is a bundle from bitmaps
+        wxVector<wxBitmap> bitmaps;
+        wxArrayString pathes = wxSplit(paramValue, ';');
```
Comment (vadz) [inline src/xrc/xmlres.cpp:2036]: Two things again: first, there is a typo in the variable name and second, and more importantly, I don't think we want to allow escaping semicolons with backslashes here, which can be done by default, as backslashes can occur naturally in paths, at least under MSW, so:
```suggestion
        wxArrayString paths = wxSplit(paramValue, ';', '\0');
```
Alternatively, perhaps we could `';'` itself as the escape to allow escaping it inside the file names by just doubling them? If this works -- could you please test it? -- we probably should do it, and then document it.

File: src/xrc/xmlres.cpp [hunk 11]
```diff
@@ -1892,36 +1938,110 @@ wxBitmap wxXmlResourceHandlerImpl::GetBitmap(const wxXmlNode* node,
...
+        wxBitmapBundle stockArt(wxArtProvider::GetBitmapBundle(art_id, art_client, size));
+        if ( stockArt.IsOk() )
+            return stockArt;
+    }
+
+    wxBitmapBundle bitmapBundle;
+    wxString paramValue = GetParamValue(node);
+    bool isBitmaps = paramValue.Contains(";");
+    bool isSVG = paramValue.EndsWith(".svg");
+    if ( isBitmaps && isSVG )
```
Comment (vadz) [inline src/xrc/xmlres.cpp:1973]: Actually, after looking at the entire function, perhaps we don't need these variables at all. Why not write it like this:
```cpp
if ( paramValue.EndsWith(".svg") )
{
    if ( paramValue.Contains(";") )
         ... error ...
}
else // one or more ';'-separated bitmaps
{
    ...
}
```
?

File: src/xrc/xmlres.cpp [hunk 12]
```diff
@@ -1892,36 +1938,110 @@ wxBitmap wxXmlResourceHandlerImpl::GetBitmap(const wxXmlNode* node,
...
+        }
+    }
+    else
+    {
+        // it is a bundle from bitmaps
+        wxVector<wxBitmap> bitmaps;
+        wxArrayString pathes = wxSplit(paramValue, ';');
+        for ( wxArrayString::const_iterator i = pathes.begin(); i != pathes.end(); ++i )
+        {
+            bitmaps.push_back(LoadBitmapFromFile(this, *i, size, param));
```
Comment (vadz) [inline src/xrc/xmlres.cpp:2039]: We need to check for errors here, at the very least we shouldn't add invalid bitmaps to the vector as this would result in asserts from `wxBitmapBundle::FromBitmaps()`. I'm not sure if we should return a valid bundle or an invalid one if any of the bitmaps fail to load, but we definitely must return an invalid bundle without asserting if all of them fail to load.

**Review by vadz — COMMENTED** [review]
It would be nice if you could please update the xrc sample to use a non-trivial (i.e. containing more than one bitmap) bundle somewhere. TIA!

**Review by vadz — APPROVED** [review]
Looks good to me, thanks!

If there are no objections, I'm going to merge this soon.

**Comment by ousnius** [comment]
@vadz @kosh543 This merge indirectly caused a regression with the expansion of environment variables in XRC bitmap file paths for wxButton (added with https://github.com/wxWidgets/wxWidgets/pull/1445).

``GetBitmapBundle`` is missing calls to ``GetFilePath`` or similar functionality like ``GetBitmap`` is doing, which includes the expansion of env vars in paths.
